torch/nn/modules/module.py

```diff
@@ -340,21 +340,75 @@ def parameters(self, memo=None):
                 yield p
 
     def children(self):
-        """Returns an iterator over children modules."""
+        """Returns an iterator over immediate children modules."""
+        for name, module in self.named_children():
+            yield module
+
+    def named_children(self):
+        """Returns an iterator over immediate children modules, yielding both
+        the name of the module as well as the module itself.
+
+        Example:
+            >>> for name, module in model.named_children():
+            >>>     if name in ['conv4', 'conv5']:
+            >>>         print(module)
+        """
         memo = set()
-        for module in self._modules.values():
+        for name, module in self._modules.items():
             if module is not None and module not in memo:
                 memo.add(module)
-                yield module
+                yield name, module
+
+    def modules(self):
+        """Returns an iterator over all modules in the network.
+
+        Note:
+            Duplicate modules are returned only once. In the following
+            example, ``l`` will be returned only once.
+
+            >>> l = nn.Linear(2, 2)
+            >>> net = nn.Sequential(l, l)
+            >>> for idx, m in enumerate(net.modules()):
+            >>>     print(idx, '->', m)
+            0 -> Sequential (
+              (0): Linear (2 -> 2)
+              (1): Linear (2 -> 2)
+            )
+            1 -> Linear (2 -> 2)
+        """
+        for name, module in self.named_modules():
+            yield module
+
+    def named_modules(self, memo=None, prefix=''):
+        """Returns an iterator over all modules in the network, yielding
+        both the name of the module as well as the module itself.
+
+        Note:
+            Duplicate modules are returned only once. In the following
+            example, ``l`` will be returned only once.
+
+            >>> l = nn.Linear(2, 2)
+            >>> net = nn.Sequential(l, l)
+            >>> for idx, m in enumerate(net.named_modules()):
+            >>>     print(idx, '->', m)
+            0 -> ('', Sequential (
+              (0): Linear (2 -> 2)
+              (1): Linear (2 -> 2)
+            ))
+            1 -> ('0', Linear (2 -> 2))
+        """
 
-    def modules(self, memo=None):
         if memo is None:
             memo = set()
         if self not in memo:
             memo.add(self)
-            yield self
-            for module in self.children():
-                for m in module.modules(memo):
+            yield prefix, self
+            for name, module in self.named_children():
+                if prefix:
+                    submodule_prefix = prefix + '.' + name
+                else:
+                    submodule_prefix = name
```

Comment:
Didn't manage to make it work without a conditional on `prefix` being an empty string. Feel free to improve it